src/nccl_ofi_rdma.c

```diff
@@ -3365,6 +3409,74 @@ static int send_comm_destroy(nccl_net_ofi_rdma_send_comm_t *s_comm)
 	return ret;
 }
 
+/**
+ * Process all finalizing s_comm's, and destroy if done
+ *
+ * Note: caller must own the comm_cleanup_list_lock when calling
+ * this function
+ */
+static int send_comm_process_all_finalizing(void)
+{
+	int ret = 0;
+
+	nccl_net_ofi_rdma_send_comm_t *s_comm = NULL, *tmp = NULL;
+	DL_FOREACH_SAFE(s_comm_cleanup_list, s_comm, tmp) {
```

Comment:
Calling this out. During the design review @bwbarrett didn't want us to iterate the cleanup list here. But I think it's unavoidable: the communicators can have different completion queues (`cq` is thread local), and we need to make progress on all of them.

Reply:
we have a dequeue; don't use UTHash's list here.  Use our internal code so we have less crap to cleanup later.

I'm not thrilled with the polling, but you're right that there's not an obvious way around it.

Reply:
> we have a dequeue; don't use UTHash's list here. Use our internal code so we have less crap to cleanup later.

Would like to understand why you prefer using our internal deque here; I don't see the advantage. Note that:

1. Our deque is missing functionality we need here (removing an element in the middle, and iterating), which we would have to implement and test
2. We already use utlist in the endpoint-per-comm code for reason (1), so this isn't the first use of utlist in the code.

Once (if) we move to C++, we get deque functionality for free, and we can eliminate both utlist and our internal deque then. In the mean time, I might argue we shouldn't spend time adding functionality to our internal deque.

Reply:
We shouldn't have two different list structures we're using in the code.  I missed the endpoint-per-comm usage, we should fix that.  Please open a SIM.

Reply:
> We shouldn't have two different list structures we're using in the code.

Updated this PR to use `nccl_ofi_deque_t` and implemented missing functionality in the deque.

> I missed the endpoint-per-comm usage, we should fix that. Please open a SIM.

Opened.

Reply:
If you folks review https://github.com/aws/aws-ofi-nccl/pull/557 you can rename this file to `nccl_ofi_rdma.cpp` and just use std::vector?

Reply:
After thinking about it, I'm not sure C++'s built-in deque actually does exactly what we want. When we wrote our internal deque, we sought to have a deque without memory allocation in the critical path (storing list data within the contained type itself). C++'s deque would require memory allocation, I believe? (Unless we write a custom allocator?)

In any case, I'm guessing the switch to C++ will come after this PR lands.

Reply:
You would use https://en.cppreference.com/w/cpp/memory/unsynchronized_pool_resource